collector/framework_collector.go

```diff
@@ -0,0 +1,302 @@
+			continue
+		}
+		tcpI("Connection accepted.")
+		tcpI(fmt.Sprintf("Launching handler for TCP connection from: %+v", conn.RemoteAddr()))
+		go handleConnection(conn, recordsChan)
+	}
+}
+
+// ---
+
+// Function which reads records from a TCP session.
+// This function should be run as a gofunc.
+func handleConnection(conn net.Conn, recordsChan chan<- *AvroDatum) {
+	defer conn.Close()
+	reader := &countingReader{conn, 0}
+	avroReader, err := goavro.NewReader(goavro.FromReader(reader))
+	if err != nil {
+		tcpE(fmt.Sprintf("Failed to create avro reader: %s", err))
+		return // close connection
+	}
+	defer func() {
+		if err := avroReader.Close(); err != nil {
+			tcpE(fmt.Sprintf("Failed to close avro reader:", err))
+		}
+	}()
+
+	nextInputResetTime := time.Now().Add(time.Second * time.Duration(inputLimitPeriodFlag))
+	var lastBytesCount int64
+	var recordCount int64
+	for {
+		lastBytesCount = reader.inputBytes
+		// Wait for records to be available:
+		if !avroReader.Scan() {
+			// Stream closed, exit
+			break
+		}
+		datum, err := avroReader.Read()
+		if err != nil {
+			log.Errorf("Cannot read avro record from %+v: %s\n", conn.RemoteAddr(), err)
+			continue
+		}
+
+		// TODO(roger): !ok
+		topic, _ := GetTopic(datum)
+
+		// increment counters before reader.inputBytes is modified too much
+		// NOTE: inputBytes is effectively being modified by a gofunc in avroReader, so it's not a perfect measurement
+		recordCount++
+		approxBytesRead := reader.inputBytes - lastBytesCount
+
+		// reset throttle counter if needed, before enforcing it below
+		// ideally we'd use a ticker for this, but the goavro api already requires we use manual polling
+		now := time.Now()
+		if now.After(nextInputResetTime) {
+			// Limit period has transpired, reset limit count before continuing
+			// TODO(MALNICK) Don't case like this...
+			if reader.inputBytes > int64(inputLimitAmountKBytesFlag)*1024 {
+				log.Debugf("INPUT SUMMARY: Received %d MetricLists (%d KB) from %s in the last ~%ds. "+
+					"Of this, ~%d KB was dropped due to throttling.\n",
+					recordCount,
+					reader.inputBytes/1024,
+					conn.RemoteAddr(),
+					inputLimitPeriodFlag,
+					// TODO (MALNICK) don't cast like this..
+					reader.inputBytes/1024-int64(inputLimitAmountKBytesFlag))
+			} else {
+				log.Infof("INPUT SUMMARY: Received %d MetricLists (%d KB) from %s in the last ~%ds\n",
+					recordCount, reader.inputBytes/1024, conn.RemoteAddr(), inputLimitPeriodFlag)
+			}
+			recordCount = 0
+			reader.inputBytes = 0
+			nextInputResetTime = now.Add(time.Second * time.Duration(inputLimitPeriodFlag))
+		}
+
+		// TODO (MALNICK) Don't cast like this
+		if reader.inputBytes > int64(inputLimitAmountKBytesFlag)*1024 {
+			// input limit reached, skip
+			continue
+		}
+		recordsChan <- &AvroDatum{datum, topic, approxBytesRead}
+	}
+}
+
+func GetTopic(obj interface{}) (string, bool) {
+	record, ok := obj.(*goavro.Record)
+	if !ok {
+		return "UNKNOWN_RECORD_TYPE", false
```

Comment:
can we return error here instead of a string and handle the error in the caller?